File: services/user-service/legacy-node/src/middleware/jwtMiddleware.js
// Same middleware logic, identical to auth-service version.
// Each service has its own copy so services are independently deployable.
const jwt = require('jsonwebtoken');
const fs  = require('fs');

const publicKey = fs.readFileSync(process.env.JWT_PUBLIC_KEY_PATH || './keys/public.pem');

function verifyJWT(req, res, next) {
  const authHeader = req.headers.authorization || '';
  const token = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : null;
  if (!token) return res.status(401).json({ error: 'Missing token' });

  try {
    const payload = jwt.verify(token, publicKey, { algorithms: ['RS256'] });
    req.user = payload;
    next();
  } catch (err) {
    const message = err.name === 'TokenExpiredError' ? 'Token expired' : 'Invalid token';
    return res.status(401).json({ error: message });
  }
}

const ROLE_RANK = { user: 0, admin: 1 };

function requireRole(minRole) {
  return (req, res, next) => {
    const userRank = ROLE_RANK[req.user?.role] ?? -1;
    const minRank  = ROLE_RANK[minRole] ?? 99;
    if (userRank < minRank) return res.status(403).json({ error: 'Insufficient permissions' });
    next();
  };
}

module.exports = { verifyJWT, requireRole };

File: services/user-service/internal/middleware/jwt_middleware.go
package middleware

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

type JWTMiddleware struct{}

func NewJWTMiddleware(path string) (*JWTMiddleware, error) {
	return &JWTMiddleware{}, nil
}

func (m *JWTMiddleware) VerifyJWT() gin.HandlerFunc {
	return func(c *gin.Context) {
		// ADR-002: gateway validates JWT and forwards trusted identity headers.
		sub := c.GetHeader("X-User-Id")
		role := c.GetHeader("X-User-Role")
		email := c.GetHeader("X-User-Email")
		if sub == "" || role == "" {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Missing token"})
			return
		}
		c.Set("user", map[string]any{"sub": sub, "role": role, "email": email})
		c.Next()
	}
}

func RequireRole(minRole string) gin.HandlerFunc {
	roleRank := map[string]int{"user": 0, "admin": 1}
	return func(c *gin.Context) {
		v, ok := c.Get("user")
		if !ok {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Missing token"})
			return
		}
		claims := v.(map[string]any)
		role, _ := claims["role"].(string)
		if roleRank[role] < roleRank[minRole] {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{"error": "Insufficient permissions"})
			return
		}
		c.Next()
	}
}
